ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/FunctionArgumentsUtil.java

```diff
@@ -0,0 +1,300 @@
+        final SqlArgument sqlArgument,
+        final TypeContext context
+    ) {
+      this.sqlArgument = sqlArgument;
+      this.context = context;
+    }
+
+    public SqlArgument getSqlArgument() {
+      return sqlArgument;
+    }
+
+    public TypeContext getContext() {
+      return context;
+    }
+  }
+
+  /**
+   * Compute type information given a function call node. Specifically, computes
+   * the function return type, the types of all arguments, and the types of all
+   * lambda parameters for arguments that are lambda expressions.
+   *
+   * <p>Given a function call node, we have to do a two pass processing of the
+   * function arguments in order to properly handle any potential lambda functions.</p>
+   *
+   * <p>In the first pass, if there are lambda functions, we create a SqlLambda that only contains
+   * the number of input arguments for the lambda. We pass this first argument list
+   * to UdfFactory in order to get the correct function. We can make this assumption 
+   * due to Java's handling of type erasure (Function(T,R) is considered the same as
+   * Function(U,R)).</p>
+   *
+   * <p>In the second pass, we use the LambdaType inputTypes field to construct SqlLambdaResolved
+   * that has the proper input type list and return type. We also need to construct a list of
+   * contexts that should be used when processing each function argument subtree.</p>
+   *
+   * @param expressionTypeManager an expression type manager
+   * @param functionCall  the function expression
+   * @param udfFactory  udf factory for the function in the expression
+   * @param context a type context
+   *
+   * @return a wrapper that contains a list of function arguments 
+   *         (any lambdas are SqlLambdaResolved), the ksql function, 
+   *         type contexts for use in further processing the function 
+   *         argument child nodes, and the return type of the ksql function
+   */
+  // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
+  public static FunctionTypeInfo getFunctionTypeInfo(
+      final ExpressionTypeManager expressionTypeManager,
+      final FunctionCall functionCall,
+      final UdfFactory udfFactory,
+      final TypeContext context
+  ) {
+    // CHECKSTYLE_RULES.ON: CyclomaticComplexity
+    final List<Expression> arguments = functionCall.getArguments();
+    final List<SqlArgument> functionArgumentTypes = firstPassOverFunctionArguments(
+        arguments,
+        expressionTypeManager,
+        context
+    );
+
+    final KsqlScalarFunction function = udfFactory.getFunction(functionArgumentTypes);
+    final SqlType returnSchema;
+    final List<TypeContext> typeContextsForChildren = new ArrayList<>();
+
+    if (!functionCall.hasLambdaFunctionCallArguments()) {
+      returnSchema = function.getReturnType(functionArgumentTypes);
+      return FunctionTypeInfo.of(
+          functionArgumentTypes.stream()
+              .map(argument -> ArgumentInfo.of(argument, context.getCopy()))
```

Comment:
Just wondering: could we reuse the copied context from `firstPassOverFunctionArguments` since that one seems not modified till here?

Reply:
All the contexts need to be separate instances when processing child nodes of the function call arguments so this seems like the easiest way to do it.